qcodes/instrument/abstract.py

```diff
@@ -0,0 +1,96 @@
+    raise this exception class.
+    """
+
+
+def abstract_instrument(klass: Type[InstrumentBase]) -> Type[InstrumentBase]:
+    """
+    A class decorator to create an abstract instrument. Abstract
+    instruments are allowed to have abstract parameters, but
+    their subclasses do not. This works by replacing the
+    'add_parameter' and '__init_subclass__' methods of the class
+    dynamically.
+
+    Args:
+        klass: The class to be decorated
+
+    Returns:
+        The decorated class
+    """
+    _instrument_base_type: Any = klass  # see https://github.com/python/mypy/issues/5865
+
+    class Abstract(_instrument_base_type):
+        def __init_subclass__(cls: Type[InstrumentBase]) -> None:
+
+            __init__ = cls.__init__
+
+            def __init_new__(self: InstrumentBase, *args: Any, **kwargs: Any) -> None:
+                """
+                Subclasses of an abstract instrument should check
+                after initialization whether there still are
+                abstract parameters. If there are, we should raise
+                an exception.
+                """
+                __init__(self, *args, **kwargs)
+                # after the usual initialization...
+
+                abstract_parameters = [
+                    parameter.name for parameter in self.parameters.values()
+                    if isinstance(parameter, AbstractParameter)
+                ]
+
+                if any(abstract_parameters):
+                    cls_name = cls.__name__
+
+                    raise AbstractParameterException(
+                        f"Class '{cls_name}' has un-implemented Abstract Parameter(s): "
+                        f"" + ", ".join([f"'{name}'" for name in abstract_parameters])
+                    )
+
+            # See https://github.com/python/mypy/issues/2427
+            cls.__init__ = __init_new__  # type: ignore
+
+        def add_parameter(
+                self, name: str, parameter_class: Type[_BaseParameter] = Parameter,
+                **kwargs: Any
+        ) -> None:
+
+            existing_parameter = self.parameters.get(name, None)
+
+            if isinstance(existing_parameter, AbstractParameter):
+                # For abstract parameters, we define special behavior.
+                existing_unit = getattr(existing_parameter, "unit", None)
+                new_unit = kwargs.get("unit", None)
+
+                if existing_unit and existing_unit != new_unit:
+                    raise AbstractParameterException(
+                        f"The unit of the parameter '{name}' is '{new_unit}', "
+                        f"which is inconsistent with the unit '{existing_unit}' "
+                        f"specified in the baseclass {klass.__name__!r} "
+                    )
+
+                # Remove the original abstract parameter to make room for the
+                # implementation in the subclass.
+                del self.parameters[name]
+
```

Comment:
This method does not integrate very nicely with the add parameter in the base class. Deleting the parameter here seems ugly. Why not put this code inside the standard add_parameter logic such that this method can be removed completely